src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Build/OutputGroupProvider.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Build
+{
+    [Export(typeof(IOutputGroupProvider))]
+    [AppliesTo(ProjectCapabilities.VisualStudioWellKnownOutputGroups)]
+    [Order(1000)]
+    internal class OutputGroupProvider : IOutputGroupProvider
+    {
+        /// <summary>
+        /// List of well known output groups names and their associated target and description
+        /// </summary>
+        private static readonly ImmutableHashSet<IOutputGroup> s_outputGroups = ImmutableHashSet.Create<IOutputGroup>()
+            .Add(NewGroup("PublishItems", "PublishItemsOutputGroup", Resources.OutputGroupPublishItemsDisplayName, Resources.OutputGroupPublishItemsDescription));
+
+        /// <summary>
+        /// The set of output groups that apply to this project.
+        /// </summary>
+        private readonly AsyncLazy<IImmutableSet<IOutputGroup>> _outputGroups;
+
+        private readonly IProjectAccessor _projectAccessor;
+        private readonly ConfiguredProject _configuredProject;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OutputGroupProvider"/> class.
+        /// </summary>
+        /// <param name="projectAccessor">Imported <see cref="IProjectAccessor"/>.</param>
+        /// <param name="configuredProject">Imported <see cref="ConfiguredProject"/>.</param>
+        [ImportingConstructor]
+        private OutputGroupProvider(IProjectAccessor projectAccessor, ConfiguredProject configuredProject)
+        {
+            _projectAccessor = projectAccessor;
+            _configuredProject = configuredProject;
+            _outputGroups = new AsyncLazy<IImmutableSet<IOutputGroup>>(GetOutputGroupMetadataAsync);
```

Comment:
This looks wrong; this will fix the result for the entire lifetime of the project even though targets can be removed and added. Is this what CPS does?

Reply:
Yes, this is what CPS does.  This list of targets is used for filtering out items from s_outputGroups that may not be applicable to this project.  I don't think there's a scenario where the PublishItemsOutputGroup target would get added/removed after the project is loaded.

Reply:
>  I don't think there's a scenario where the PublishItemsOutputGroup target would get added/removed after the project is loaded.

Given CPS already does this at the moment, I'm okay with this, but absolutely this could occur via a simple branch switch; project starts off as say a legacy project (that we've automatically opened in CPS - coming soon) and you switch to a branch where it is an SDK-based project. The project stays loaded and we apply the diff.